entry_proxy/main.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+/*  Entry Proxy for Onion Gateway
+
+See also:
+
+  * https://github.com/DonnchaC/oniongateway/blob/master/docs/design.rst#32-entry-proxy
+  * https://gist.github.com/Yawning/bac58e08a05fc378a8cc (SOCKS5 client, Tor)
+  * https://habrahabr.ru/post/142527/ (TCP proxy)
+*/
+
+import (
+    "flag"
+    "log"
+    "net"
+    "net/url"
+    "strconv"
+    "time"
+
+    "github.com/polvi/sni"
+    "golang.org/x/net/proxy"
+)
+
+var (
+    proxyUrl = flag.String("proxy", "socks5://127.0.0.1:9050", "Proxy URL")
+    listenOn = flag.String("listen-on", "0.0.0.0:443", "Where to listen")
+    onionPort = flag.Int("onion-port", 443, "Port on onion site to use")
+    bufferSize = flag.Int("buffer-size", 1024, "Proxy buffer size, bytes")
+    timeout = flag.Duration("timeout", 60 * time.Second, "Timeout for IO")
+)
+
+func connectToProxy(targetServer string) (net.Conn, error) {
+    parsedUrl, err := url.Parse(*proxyUrl)
+    if err != nil {
+        return nil, err
+    }
+    dialer, err := proxy.FromURL(parsedUrl, proxy.Direct)
+    if err != nil {
+        return nil, err
+    }
+    connection, err := dialer.Dial("tcp", targetServer)
+    return connection, err
+}
+
+func netCopy(from, to net.Conn, finished chan<- struct{}) {
+    defer func() {
+        finished<-struct{}{}
+    }()
+    buffer := make([]byte, *bufferSize)
+    for {
+        from.SetReadDeadline(time.Now().Add(*timeout))
```

Comment:
It works without setting the deadline. The timeout was introduced to 
deal with peers who do not close the connection properly. They can 
exhaust all sockets and stop normal operation of the service. 
However this code was removed, because we have to confirm this 
problem before fixing it.